Fix #2134: Fix concept card naming/linking

BenHenning opened the pull request:
Fix #2134 

This fixes an issue with HtmlParser where we weren't correctly unescaping the concept card name/link. Note that this inconsistency may have arisen from the shift to using CustomHtmlContentHandler which partially parses the JSON-escaped custom attribute values (which turns ``&amp;&quot;`` to just ``&quot;``).

This PR also consolidates all custom tags into using the CustomHtmlContentHandler to avoid global unescaping (e.g. replacing all ``&amp;&quot;`` with empty string) since that's incorrect and breaks encoded JSON objects (one of which is used for math tags).

There are also some minor refactors/changes occurring here:
- The concept card tag handler is being split into its own file
- A new handler is being introduced to handle image handling to address the inconsistency above
- Tests are added for both the new handlers & for HtmlParserTest (there are already higher-level tests for concept cards in StateFragmentTest)
- The Mockito helper was moved from app module to testing so that it can be used in the new utility tests
- This PR is preparing for more complex downstream image loading (see #2942) by introducing an ImageRetriever interface for CustomHtmlContentHandler that allows custom tag handlers to specify how images should be styled (only the current 'block' format is supported; another will be added in a later PR for in-line images)
- There are fixes added to HtmlParserTest to make it more correct: it could previously interact with UI objects off the main thread (this was discovered when trying to run the updated tests with Espresso). Some helper methods were added to perform operations on the main thread with issues.

BenHenning committed:
Add support for loading proto versions of lessons.

BenHenning committed:
Upgrade Espresso versions to work around build issues.

Note that I can't actually verify Espresso tests are working with these
versions since the Bazel project doesn't yet have Espresso tests set up.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Lint fix post-copying from #2927.

BenHenning committed:
Fix Glide in Bazel.

BenHenning committed:
Fix concept card name escaping.

Also, refactor HtmlParser to better support/handling different tag types
& prepare for LaTeX SVG image support.

BenHenning committed:
Copy over test fix from #2927.

BenHenning committed:
Remove image loading annotation.

This will be added in a later branch in the chain.

BenHenning committed:
Add caching module to needed test suites.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning commented:
/cc @rt4914

rt4914 commented on utility/src/main/java/org/oppia/android/util/parser/ConceptCardTagHandler.kt:
```diff
@@ -0,0 +1,42 @@
+            customOppiaTagActionListener?.onConceptCardLinkClicked(view, skillId)
+          }
+        },
+        0, text.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE
```
Suggest adding inline comment
`/* start= */ 0, text.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE`

BenHenning replied:
Nice catch--fixed.

rt4914 commented on utility/src/main/java/org/oppia/android/util/parser/ImageTagHandler.kt:
```diff
@@ -0,0 +1,48 @@
+import org.oppia.android.util.logging.ConsoleLogger
+import org.xml.sax.Attributes
+
+/** The custom tag corresponding to [MathTagHandler]. */
```
`MathTagHandler ` file does not exists.

BenHenning replied:
Thanks--fixed.

rt4914 reviewed (commented):
Overall implementation looks good. Thanks.

BenHenning committed:
Add support for importing text protos as assets.

This actually provides Bazel-only support for converting text protos to
binary and including those as assets so that we can avoid checking in
binary protos to the codebase.

BenHenning committed:
Add gitignore for Android Studio Bazel plugin output.

BenHenning committed:
Import textproto versions of existing dev assets.

This imports local conversions of all existing dev assets as text proto
leveraging the Bazel-only conversion system to convert these to binary
protos and include them as assets.

This is a necessary step to both eventually replace JSON assets with
text proto, and to test that the new proto loading pipeline is working
as designed.

BenHenning committed:
Undo Espresso version change since it breaks Bazel tests.

BenHenning committed:
Fix tests.

Various attempts to fix existing tests + move some of the domain tests
over to actually using proto when running with Bazel.

App module tests are hanging currently for unknown reasons (only in
Bazel).

BenHenning committed:
Use correct Espresso core.

3.1.0 causes tests to hang indefinitely when run with Bazel.

BenHenning committed:
Move coroutine dispatchers.

This moves the coroutine dispatchers to a new threading subpackage. It
also simplifies the TestCoroutineDispatcher interface & implementations
(which required moving runUntilIdle to CoroutineExecutorServiceTest).

The test coroutine annotations were also moved to their own files.

BenHenning committed:
Refactor testing utilities build graph.

This introduces new Bazel libraries for:
- Test dispatchers
- Fake system/Oppia clocks
- Robolectric dependencies

BenHenning committed:
Introduce dedicated test for TestCoroutineDispatcher.

This also fixes some threading issues in the dispatcher, and clarifies
some of its API that was previously unclear (and can lead to subtle race
conditions).

BenHenning committed:
Add tests for TestCoroutineDispatcher.

This required a bunch of refactoring, adding the first examples of
dedicated Bazel test build files, and introducing a new test pattern for
sharing code between different implementations. Further, it refined the
API for TesCoroutineDispatcher & fixed some issues in the
implementations (especially Robolectric).

Tests verified as non-flaky in Robolectric across 1000 runs, and 0.2%
flaky across 1000 runs for Espresso (though that may be fixed; waiting
on a follow-up run to confirm stability).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Workaround ktlint semicolon issue.

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix test post-merge.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Move CoroutineDispatcher to be part of threading.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into fix-protobuf-duplication-issue

BenHenning committed:
Undo version upgrade.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	.gitignore
	app/src/sharedTest/java/org/oppia/android/app/story/StoryActivityTest.kt

BenHenning committed:
Fix broken tests.

BenHenning committed:
Lint fixes.

BenHenning committed:
Add a bit more proto loading test coverage.

BenHenning committed:
Comment fix.

BenHenning committed:
Lint fixes.

BenHenning committed:
Address earlier TODO.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	tools/BUILD.bazel

BenHenning committed:
Post-merge restructure.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	data/BUILD.bazel

BenHenning committed:
Post-merge fixes.

This removes unneeded dependencies from PersistentCacheStore, including
Glide.

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Add tests for new tag handlers.

Also, add tests for HtmlParser, and clean up some existing tests.

BenHenning committed:
Fix broken test build.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into fix-glide-image-loading

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Post-merge lint fixes.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

Conflicts:
	app/BUILD.bazel
	app/src/sharedTest/java/org/oppia/android/app/parser/HtmlParserTest.kt
	testing/BUILD.bazel
	utility/BUILD.bazel

BenHenning committed:
Lint fixes.

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Post-merge fix.

BenHenning committed:
Fix typo.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning commented:
@rt4914 ready for review now.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning commented:
PTAL @anandwana001 as well since some of your codeowners files are affected.

anandwana001 reviewed (approved):
Test files LGTM

rt4914 reviewed (approved):
LGTM, Thanks.

BenHenning committed:
Merge branch 'develop' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning commented:
I plan to take a chance on Bazel for this & the next few PRs to try and get them in a bit faster. It's not ideal but I'm fairly confident the PRs won't introduce issues since Bazel tests were passing for them before.

BenHenning commented:
Also, thanks for the reviews!
